plugin/namespaces.go

```diff
@@ -0,0 +1,71 @@
+package plugin
+
+import (
+	"fmt"
+	"github.com/mitchellh/cli"
+	"reflect"
+	"strings"
+)
+
+// TODO: Everything in this file needs better naming.
```

Comment:
What exactly is the purpose of this file/namespaces? I know it hides plugin commands from the normal commands list but I don't really understand the "namespaces" part.

Reply:
Consider a plugin command like this: `trellis-one-two-three`:

When we add the `one two three` command to `mitchellh/cli`, it knows `one` and `one two` namespaces are missing and generates `MockCommand` for them. See: https://github.com/mitchellh/cli/blob/5454ffe87bc5c6d8b6b21c825617755e18a07828/cli.go#L371-L380

If we do not hide those `MockCommand`, they will show up on the command list with empty synopsis and mix with the core command list:

```sh-session
$ ./trellis-cli
Usage: trellis [--version] [--help] <command> [<args>]

Available commands are:
// ...
    new           Creates a new Trellis project
    one
    provision     Provisions the specified environment
// ...

$ ./trellis-cli one
This command is accessed by using one of the subcommands below.

Subcommands:
    two


$ ./trellis-cli one two
This command is accessed by using one of the subcommands below.

Subcommands:
    three    Third party plugin: Forward command to trellis-one-two-three
```

---

Same reason why we need to hide both `venv` and `venv hook` when only `venv hook` was added in https://github.com/roots/trellis-cli/pull/114/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R125-R131


---

Current implementation [`6f5873d`](https://github.com/roots/trellis-cli/pull/144/commits/6f5873db72b87c5b471b7f24f1ae0e5ade404c28) hides both `one` and `two`. In some cases, the help text is useless:

```sh-session
$ ./trellis-cli one
This command is accessed by using one of the subcommands below.
```

TODO: Change it to hide top level namesapce (e.g: `one`) only.

Reply:
This should be consistent with the normal commands right? So for example, if I unhide `venv` and `venv hook`, we get this (other commands omitted):

```plain
Usage: trellis [--version] [--help] <command> [<args>]

Available commands are:
    vault         Commands for Ansible Vault
    venv

Available plugins commands are:
    one two three
```

Yet the plugin shows `one two three`. The equivalent would be only showing `one` right? Basically to be consistent we should only show top-level commands?

One other inconsistency is just running the top-level command doesn't show the subcommand:

```plain
❯ PATH="$(pwd)/testing:$PATH" ./trellis-cli another
This command is accessed by using one of the subcommands below.
```

But accessing the second sub command does work:

```plain
❯ PATH="$(pwd)/testing:$PATH" ./trellis-cli one two
This command is accessed by using one of the subcommands below.

Subcommands:
    three    Third party plugin: Forward command to trellis-another-two-three
```

Reply:
Update in [`747e8d2` (#144)](https://github.com/roots/trellis-cli/pull/144/commits/747e8d20eddd9cd945c1349539e606166d0fd34a) to show top-level commands in `$ trellis --help`?

```sh-session
$ trellis --help
Usage: trellis [--version] [--help] <command> [<args>]

Available commands are:
...
    valet         Commands for Laravel Valet
    vault         Commands for Ansible Vault


Available third party plugins commands are:
    one
    plugintest
    theplugin


$ trellis one --help
This command is accessed by using one of the subcommands below.

Subcommands:
    two


$ trellis one two --help
This command is accessed by using one of the subcommands below.

Subcommands:
    three    Third party plugin: Forward command to trellis-one-two-three
```


For `$ trellis one two three --help`, it invokes the underlying plugin command, i.e: `$ trellis-one-two-three --help`

Reply:
Definition of `namespace` changed! 

Now: A namespace is the top level command.

Take these commands as examples:

```
$ trellis check
$ trellis valet link
$ trellis one two three
```

The namespaces are `check`, `valet` and `one`.

Help wanted: We need better namings.